File: ts/src/client_group_space_ops.ts
import type {
  GroupSpaceArtifactOptions,
  GroupSpaceBindOptions,
  GroupSpaceCapabilitiesOptions,
  GroupSpaceIngestOptions,
  GroupSpaceJobsOptions,
  GroupSpaceProviderAuthOptions,
  GroupSpaceProviderCredentialStatusOptions,
  GroupSpaceProviderCredentialUpdateOptions,
  GroupSpaceProviderHealthCheckOptions,
  GroupSpaceQueryOptions,
  GroupSpaceSourcesOptions,
  GroupSpaceSpacesOptions,
  GroupSpaceStatusOptions,
  GroupSpaceSyncOptions,
} from './types.js';

type ClientCall = (op: string, args?: Record<string, unknown>) => Promise<Record<string, unknown>>;

type GroupSpaceClient = {
  call: ClientCall;
};

export interface GroupSpaceOps {
  groupSpaceStatus(options: GroupSpaceStatusOptions): Promise<Record<string, unknown>>;
  groupSpaceSpaces(options: GroupSpaceSpacesOptions): Promise<Record<string, unknown>>;
  groupSpaceCapabilities(options: GroupSpaceCapabilitiesOptions): Promise<Record<string, unknown>>;
  groupSpaceBind(options: GroupSpaceBindOptions): Promise<Record<string, unknown>>;
  groupSpaceIngest(options: GroupSpaceIngestOptions): Promise<Record<string, unknown>>;
  groupSpaceQuery(options: GroupSpaceQueryOptions): Promise<Record<string, unknown>>;
  groupSpaceSources(options: GroupSpaceSourcesOptions): Promise<Record<string, unknown>>;
  groupSpaceArtifact(options: GroupSpaceArtifactOptions): Promise<Record<string, unknown>>;
  groupSpaceJobs(options: GroupSpaceJobsOptions): Promise<Record<string, unknown>>;
  groupSpaceSync(options: GroupSpaceSyncOptions): Promise<Record<string, unknown>>;
  groupSpaceProviderCredentialStatus(options?: GroupSpaceProviderCredentialStatusOptions): Promise<Record<string, unknown>>;
  groupSpaceProviderCredentialUpdate(options?: GroupSpaceProviderCredentialUpdateOptions): Promise<Record<string, unknown>>;
  groupSpaceProviderHealthCheck(options?: GroupSpaceProviderHealthCheckOptions): Promise<Record<string, unknown>>;
  groupSpaceProviderAuth(options?: GroupSpaceProviderAuthOptions): Promise<Record<string, unknown>>;
}

const groupSpaceOps: GroupSpaceOps & ThisType<GroupSpaceClient> = {
  async groupSpaceStatus(options) {
    return this.call('group_space_status', {
      group_id: options.groupId,
      provider: options.provider ?? 'notebooklm',
    });
  },

  async groupSpaceSpaces(options) {
    return this.call('group_space_spaces', {
      group_id: options.groupId,
      provider: options.provider ?? 'notebooklm',
    });
  },

  async groupSpaceCapabilities(options) {
    return this.call('group_space_capabilities', {
      group_id: options.groupId,
      provider: options.provider ?? 'notebooklm',
    });
  },

  async groupSpaceBind(options) {
    const args: Record<string, unknown> = {
      group_id: options.groupId,
      provider: options.provider ?? 'notebooklm',
      lane: options.lane,
      action: options.action ?? 'bind',
      by: options.by ?? 'user',
    };
    if (options.remoteSpaceId) args['remote_space_id'] = options.remoteSpaceId;
    return this.call('group_space_bind', args);
  },

  async groupSpaceIngest(options) {
    const args: Record<string, unknown> = {
      group_id: options.groupId,
      provider: options.provider ?? 'notebooklm',
      lane: options.lane,
      kind: options.kind ?? 'context_sync',
      by: options.by ?? 'user',
    };
    if (options.payload) args['payload'] = options.payload;
    if (options.idempotencyKey) args['idempotency_key'] = options.idempotencyKey;
    return this.call('group_space_ingest', args);
  },

  async groupSpaceQuery(options) {
    const args: Record<string, unknown> = {
      group_id: options.groupId,
      provider: options.provider ?? 'notebooklm',
      lane: options.lane,
      query: options.query,
    };
    if (options.options) args['options'] = options.options;
    return this.call('group_space_query', args);
  },

  async groupSpaceSources(options) {
    const args: Record<string, unknown> = {
      group_id: options.groupId,
      provider: options.provider ?? 'notebooklm',
      lane: options.lane,
      action: options.action ?? 'list',
      by: options.by ?? 'user',
    };
    if (options.sourceId) args['source_id'] = options.sourceId;
    if (options.newTitle) args['new_title'] = options.newTitle;
    return this.call('group_space_sources', args);
  },

  async groupSpaceArtifact(options) {
    const args: Record<string, unknown> = {
      group_id: options.groupId,
      provider: options.provider ?? 'notebooklm',
      lane: options.lane,
      action: options.action ?? 'list',
      by: options.by ?? 'user',
    };
    if (options.kind) args['kind'] = options.kind;
    if (options.options) args['options'] = options.options;
    if (options.wait !== undefined) args['wait'] = options.wait;
    if (options.saveToSpace !== undefined) args['save_to_space'] = options.saveToSpace;
    if (options.outputPath) args['output_path'] = options.outputPath;
    if (options.outputFormat) args['output_format'] = options.outputFormat;
    if (options.artifactId) args['artifact_id'] = options.artifactId;
    if (options.timeoutSeconds !== undefined) args['timeout_seconds'] = options.timeoutSeconds;
    if (options.initialInterval !== undefined) args['initial_interval'] = options.initialInterval;
    if (options.maxInterval !== undefined) args['max_interval'] = options.maxInterval;
    return this.call('group_space_artifact', args);
  },

  async groupSpaceJobs(options) {
    const args: Record<string, unknown> = {
      group_id: options.groupId,
      provider: options.provider ?? 'notebooklm',
      lane: options.lane,
      action: options.action ?? 'list',
      by: options.by ?? 'user',
    };
    if (options.jobId) args['job_id'] = options.jobId;
    if (options.state) args['state'] = options.state;
    if (options.limit !== undefined) args['limit'] = options.limit;
    return this.call('group_space_jobs', args);
  },

  async groupSpaceSync(options) {
    return this.call('group_space_sync', {
      group_id: options.groupId,
      provider: options.provider ?? 'notebooklm',
      lane: options.lane,
      action: options.action ?? 'status',
      force: options.force ?? false,
      by: options.by ?? 'user',
    });
  },

  async groupSpaceProviderCredentialStatus(options = {}) {
    return this.call('group_space_provider_credential_status', {
      provider: options.provider ?? 'notebooklm',
      by: options.by ?? 'user',
    });
  },

  async groupSpaceProviderCredentialUpdate(options = {}) {
    const args: Record<string, unknown> = {
      provider: options.provider ?? 'notebooklm',
      by: options.by ?? 'user',
      clear: options.clear ?? false,
    };
    if (options.authJson) args['auth_json'] = options.authJson;
    return this.call('group_space_provider_credential_update', args);
  },

  async groupSpaceProviderHealthCheck(options = {}) {
    return this.call('group_space_provider_health_check', {
      provider: options.provider ?? 'notebooklm',
      by: options.by ?? 'user',
    });
  },

  async groupSpaceProviderAuth(options = {}) {
    const args: Record<string, unknown> = {
      provider: options.provider ?? 'notebooklm',
      action: options.action ?? 'status',
      by: options.by ?? 'user',
    };
    if (options.timeoutSeconds !== undefined) args['timeout_seconds'] = options.timeoutSeconds;
    return this.call('group_space_provider_auth', args);
  },
};

export function installGroupSpaceOps(proto: GroupSpaceClient & Partial<GroupSpaceOps>): void {
  Object.assign(proto, groupSpaceOps);
}

File: python/src/cccc_sdk/client_group_space_ops.py
from __future__ import annotations

from typing import Any, Dict, Optional


class GroupSpaceOpsMixin:
    def group_space_status(self, *, group_id: str, provider: str = "notebooklm") -> Dict[str, Any]:
        return self.call("group_space_status", {"group_id": str(group_id), "provider": str(provider)})

    def group_space_spaces(self, *, group_id: str, provider: str = "notebooklm") -> Dict[str, Any]:
        return self.call("group_space_spaces", {"group_id": str(group_id), "provider": str(provider)})

    def group_space_capabilities(self, *, group_id: str, provider: str = "notebooklm") -> Dict[str, Any]:
        return self.call("group_space_capabilities", {"group_id": str(group_id), "provider": str(provider)})

    def group_space_bind(
        self,
        *,
        group_id: str,
        lane: str,
        action: str = "bind",
        remote_space_id: str = "",
        provider: str = "notebooklm",
        by: str = "user",
    ) -> Dict[str, Any]:
        args: Dict[str, Any] = {
            "group_id": str(group_id),
            "provider": str(provider),
            "lane": str(lane),
            "action": str(action),
            "by": str(by),
        }
        if remote_space_id:
            args["remote_space_id"] = str(remote_space_id)
        return self.call("group_space_bind", args)

    def group_space_ingest(
        self,
        *,
        group_id: str,
        lane: str,
        kind: str = "context_sync",
        payload: Optional[Dict[str, Any]] = None,
        idempotency_key: str = "",
        provider: str = "notebooklm",
        by: str = "user",
    ) -> Dict[str, Any]:
        args: Dict[str, Any] = {
            "group_id": str(group_id),
            "provider": str(provider),
            "lane": str(lane),
            "kind": str(kind),
            "by": str(by),
        }
        if payload is not None:
            args["payload"] = dict(payload)
        if idempotency_key:
            args["idempotency_key"] = str(idempotency_key)
        return self.call("group_space_ingest", args)

    def group_space_query(
        self,
        *,
        group_id: str,
        lane: str,
        query: str,
        options: Optional[Dict[str, Any]] = None,
        provider: str = "notebooklm",
    ) -> Dict[str, Any]:
        args: Dict[str, Any] = {
            "group_id": str(group_id),
            "provider": str(provider),
            "lane": str(lane),
            "query": str(query),
        }
        if options is not None:
            args["options"] = dict(options)
        return self.call("group_space_query", args)

    def group_space_sources(
        self,
        *,
        group_id: str,
        lane: str,
        action: str = "list",
        source_id: str = "",
        new_title: str = "",
        provider: str = "notebooklm",
        by: str = "user",
    ) -> Dict[str, Any]:
        args: Dict[str, Any] = {
            "group_id": str(group_id),
            "provider": str(provider),
            "lane": str(lane),
            "action": str(action),
            "by": str(by),
        }
        if source_id:
            args["source_id"] = str(source_id)
        if new_title:
            args["new_title"] = str(new_title)
        return self.call("group_space_sources", args)

    def group_space_artifact(
        self,
        *,
        group_id: str,
        lane: str,
        action: str = "list",
        kind: str = "",
        options: Optional[Dict[str, Any]] = None,
        wait: Optional[bool] = None,
        save_to_space: Optional[bool] = None,
        output_path: str = "",
        output_format: str = "",
        artifact_id: str = "",
        timeout_seconds: Optional[int] = None,
        initial_interval: Optional[int] = None,
        max_interval: Optional[int] = None,
        provider: str = "notebooklm",
        by: str = "user",
    ) -> Dict[str, Any]:
        args: Dict[str, Any] = {
            "group_id": str(group_id),
            "provider": str(provider),
            "lane": str(lane),
            "action": str(action),
            "by": str(by),
        }
        if kind:
            args["kind"] = str(kind)
        if options is not None:
            args["options"] = dict(options)
        if wait is not None:
            args["wait"] = bool(wait)
        if save_to_space is not None:
            args["save_to_space"] = bool(save_to_space)
        if output_path:
            args["output_path"] = str(output_path)
        if output_format:
            args["output_format"] = str(output_format)
        if artifact_id:
            args["artifact_id"] = str(artifact_id)
        if timeout_seconds is not None:
            args["timeout_seconds"] = int(timeout_seconds)
        if initial_interval is not None:
            args["initial_interval"] = int(initial_interval)
        if max_interval is not None:
            args["max_interval"] = int(max_interval)
        return self.call("group_space_artifact", args)

    def group_space_jobs(
        self,
        *,
        group_id: str,
        lane: str,
        action: str = "list",
        job_id: str = "",
        state: str = "",
        limit: Optional[int] = None,
        provider: str = "notebooklm",
        by: str = "user",
    ) -> Dict[str, Any]:
        args: Dict[str, Any] = {
            "group_id": str(group_id),
            "provider": str(provider),
            "lane": str(lane),
            "action": str(action),
            "by": str(by),
        }
        if job_id:
            args["job_id"] = str(job_id)
        if state:
            args["state"] = str(state)
        if limit is not None:
            args["limit"] = int(limit)
        return self.call("group_space_jobs", args)

    def group_space_sync(
        self,
        *,
        group_id: str,
        lane: str,
        provider: str = "notebooklm",
        by: str = "user",
    ) -> Dict[str, Any]:
        return self.call(
            "group_space_sync",
            {
                "group_id": str(group_id),
                "provider": str(provider),
                "lane": str(lane),
                "action": "status",
                "by": str(by),
            },
        )
